python/src/aac/lang/definitions/structure.py

```diff
@@ -0,0 +1,110 @@
+
+from aac.lang.definitions.arrays import is_array_type
+from aac.lang.definitions.definition import Definition
+from aac.lang.definitions.schema import get_definition_schema, get_schema_defined_fields
+from aac.lang.language_context import LanguageContext
+
+
+def get_substructures_by_type(
+    source_definition: Definition, sub_schema_definition: Definition, context: LanguageContext
+) -> list[dict]:
+    """
+    Return a list of dictionaries that represent instances of the sub-definition type within the source definition.
+
+    For example, if a definition of type `model` where to be searched for every instance of the `Field` type,
+    then every instance of a `Field` definition within the `model` definition will be returned. This includes embedded
+    fields, such as any instances of `Field` the might be defined in the `model` under the `Behavior` definition.
+
+    Args:
+        source_definition (Definition): The definition to search through
+        sub_schema_definition (Definition): Defines the target schema to extract from the source definition
+        context (LanguageContext): The language context, used to navigate the structure and lookup definitions
+
+    Returns:
+        A list of dictionaries that match instances of the sub-definition schema.
+    """
+    substructure_instances = []
+
+    def _get_substructures(schema_definition: Definition, definition_dict: dict):
+        if schema_definition.name == sub_schema_definition.name:
+            substructure_instances.append(definition_dict)
+
+        schema_defined_fields = schema_definition.get_fields().get("fields")
+        schema_defined_fields_dict = {field.get("name"): field for field in schema_defined_fields}
+        field_names_to_traverse = set(schema_defined_fields_dict.keys()).intersection(set(definition_dict.keys()))
+
+        for field_name in sorted(field_names_to_traverse):
+            field_type = schema_defined_fields_dict.get(field_name).get("type")
+
+            if context.is_definition_type(field_type):
+                field_type_definition = context.get_definition_by_name(field_type)
+                field_content = definition_dict.get(field_name)
+
+                # Skip empty fields and enums
+                if not field_content or field_type_definition.is_enum():
+                    continue
+                # If the field type isn't an array type, package it as a one element list.
+                if not is_array_type(field_type):
+                    field_content = [field_content]
+                elif not type(field_content) is list:
+                    logging.error(
+                        f"Value '{field_content}' is not an array type like its defined type '{field_type}'. Bad value in the field '{field_name}' in definition '{source_definition.name}'."
+                    )
+                    continue
+
+                for field in field_content:
+                    _get_substructures(field_type_definition, field)
```

Comment:
You could determine that by keeping track of the definitions you've traversed so far since you always start from the top-level fields of a definition. I'd have to add some additional logic to support that -- why do you ask?

Reply:
I was asking because I think that would make #252 a bit more straightforward. Essentially, I was curious if I'd be able to leverage this, for solving that.

Reply:
I think you'll need to create a separate function that effectively performs a look-up on a substructure by traversing the definition and identifying the target substructure. That traversal/lookup will look a lot like this function, but just different enough in scope and responsibilities to deserve its own public function.

I'm having trouble envisioning the problem without understanding what information VSCode will provide to that handler, but if we can access the entire definition then we can leverage keys like we do in search to narrow the traversal.